### PR TITLE
Improve performance of BackendSamplerV2 and EstimatorV2

### LINKED ISSUE #12290: Performance regression of `BackendSamplerV2` and `BackendEstimatorV2`
### What should we add?

We updated Primitives V2 to enable SamplerV2 and EstimatorV2 to handle different numbers of shots or precision for each pub.
To address this, I implemented BackendSamplerV2 and BackendEstimatorV2 to call backend.run for each pub.
On the other hand, BackendEstimatorV1 and BackendSamplerV1 assume the same number of shots per run call and pass all circuits to backend.run only once.

Copied from https://github.com/Qiskit/qiskit-ibm-runtime/issues/1631

Qiskit main (71753689)
```python
from timeit import timeit

from qiskit_aer import AerSimulator
from qiskit_ibm_runtime.fake_provider import FakeSherbrooke

from qiskit import QuantumCircuit
from qiskit.primitives import BackendSampler, BackendSamplerV2
from qiskit.transpiler.preset_passmanagers import generate_preset_pass_manager

backend = AerSimulator.from_backend(FakeSherbrooke())
shots = 10000
num_copies = 20


def gen_circuit(num_qubits: int, reps: int):
    qc = QuantumCircuit(num_qubits)
    for _ in range(reps):
        qc.h(range(num_qubits))
        for i in range(0, num_qubits - 1, 2):
            qc.cx(i, i + 1)
    qc.measure_all()
    return qc


def bench_sampler_v1(qc: QuantumCircuit):
    print("\nBackendSamplerV1")
    sampler = BackendSampler(backend)
    print(f"{timeit(lambda: sampler.run([qc] * num_copies, shots=shots).result(), number=1)} sec")


def bench_sampler_v2(qc: QuantumCircuit):
    print("\nBackendSamplerV2")
    sampler = BackendSamplerV2(backend=backend)
    print(f"{timeit(lambda: sampler.run([qc] * num_copies, shots=shots).result(), number=1)} sec")


qc = gen_circuit(5, 5)
pm = generate_preset_pass_manager(optimization_level=2, backend=backend)
qc2 = pm.run(qc)
bench_sampler_v1(qc2)
bench_sampler_v2(qc2)
```

output
```
BackendSamplerV1
0.7136987919984676 sec

BackendSamplerV2
9.465235792002204 sec
```

### PR DESCRIPTION
**Pull request by t-imamichi**
### Summary

Fixes #12290 
Addresses https://github.com/Qiskit/qiskit-ibm-runtime/issues/1631 (need to port this PR to qiskit-ibm-runtime to close it)

TODO
- [x] BackendSamplerV2
- [x] BackendEstimatorV2
- [x] reno

### Sampler
```python
from timeit import timeit

from qiskit_aer import AerSimulator
from qiskit_ibm_runtime.fake_provider import FakeSherbrooke

from qiskit import QuantumCircuit
from qiskit.primitives import BackendSampler, BackendSamplerV2
from qiskit.transpiler.preset_passmanagers import generate_preset_pass_manager

backend = AerSimulator.from_backend(FakeSherbrooke())
shots = 10000
num_copies = 20


def gen_circuit(num_qubits: int, reps: int):
    qc = QuantumCircuit(num_qubits)
    for _ in range(reps):
        qc.h(range(num_qubits))
        for i in range(0, num_qubits - 1, 2):
            qc.cx(i, i + 1)
    qc.measure_all()
    return qc


def bench_sampler_v1(qc: QuantumCircuit):
    print("\nBackendSamplerV1")
    sampler = BackendSampler(backend)
    print(f"{timeit(lambda: sampler.run([qc] * num_copies, shots=shots).result(), number=1)} sec")


def bench_sampler_v2(qc: QuantumCircuit):
    print("\nBackendSamplerV2")
    sampler = BackendSamplerV2(backend=backend)
    print(f"{timeit(lambda: sampler.run([qc] * num_copies, shots=shots).result(), number=1)} sec")


qc = gen_circuit(5, 5)
pm = generate_preset_pass_manager(optimization_level=2, backend=backend)
qc2 = pm.run(qc)
bench_sampler_v1(qc2)
bench_sampler_v2(qc2)
```

```
# main
BackendSamplerV1
0.714542875000916 sec

BackendSamplerV2
9.41767649999747 sec
```

```
# this PR
BackendSamplerV1
0.7391576250010985 sec

BackendSamplerV2
0.7517977079987759 sec
```

-----------
### Estimator
```python
from math import sqrt
from timeit import timeit

from qiskit_aer import AerSimulator
from qiskit_ibm_runtime.fake_provider import FakeSherbrooke

from qiskit import QuantumCircuit
from qiskit.primitives import BackendEstimator, BackendEstimatorV2
from qiskit.transpiler.preset_passmanagers import generate_preset_pass_manager
from qiskit.quantum_info import random_pauli_list, SparsePauliOp

backend = AerSimulator.from_backend(FakeSherbrooke())
shots = 8192
precision = sqrt(1 / shots)
num_qubits = 5
num_copies = 10
seed = 123


def gen_circuit(num_qubits: int, reps: int):
    qc = QuantumCircuit(num_qubits)
    for _ in range(reps):
        qc.h(range(num_qubits))
        for i in range(0, num_qubits - 1, 2):
            qc.cx(i, i + 1)
    return qc


def bench_estimator_v1(qc: QuantumCircuit, obs: SparsePauliOp):
    print("\nBackendEstimatorV1")
    estimator = BackendEstimator(backend)
    print(f"{timeit(lambda: estimator.run([qc] * num_copies, [obs] * num_copies, shots=shots).result(), number=1)} sec")


def bench_estimator_v2(qc: QuantumCircuit, obs: SparsePauliOp):
    print("\nBackendEstimatorV2")
    estimator = BackendEstimatorV2(backend=backend)
    print(f"{timeit(lambda: estimator.run([(qc, obs)] * num_copies, precision=precision).result(), number=1)} sec")


qc = gen_circuit(num_qubits, 5)
pm = generate_preset_pass_manager(optimization_level=2, backend=backend)
qc2 = pm.run(qc)
obs = SparsePauliOp(random_pauli_list(num_qubits, 5, phase=False)).apply_layout(qc2.layout)
bench_estimator_v1(qc2, obs)
bench_estimator_v2(qc2, obs)
```

```
# main
BackendEstimatorV1
2.3618681670050137 sec

BackendEstimatorV2
6.282726375000493 sec
```

```
# this PR
BackendEstimatorV1
2.3475185420029447 sec

BackendEstimatorV2
2.151061958000355 sec
```

### CODE REVIEW

File: qiskit/primitives/backend_estimator_v2.py [hunk 1]
```diff
@@ -144,10 +145,48 @@ def _validate_pubs(self, pubs: list[EstimatorPub]):
...
+        results = [None] * len(pubs)
+        for shots, lst in pub_dict.items():
+            # run pubs with the same number of shots at once
+            pub_results = self._run_pubs([pubs[i] for i in lst], shots)
+            # reconstruct the result of pubs
+            for i, pub_result in zip(lst, pub_results):
+                results[i] = pub_result
+        return PrimitiveResult(results)
+
+    def _run_pubs(self, pubs: list[EstimatorPub], shots: int) -> list[PubResult]:
```
Comment (ihincks) [inline qiskit/primitives/backend_estimator_v2.py:177]:
```suggestion
    def _run_pubs(self, pubs: list[EstimatorPub], shots: int) -> list[PubResult]:
        """Compute results for pubs that all require the same value of ``shots``."""
```

File: qiskit/primitives/backend_estimator_v2.py [hunk 2]
```diff
@@ -211,7 +255,7 @@ def _calc_expval_paulis(
                 expval_map[param_index, pauli.to_label()] = (expval, variance)
         return expval_map
 
-    def _preprocessing(
+    def _preprocessing_circuits(
         self, circuit: QuantumCircuit, observable: PauliList, param_index: tuple[int, ...]
     ) -> list[QuantumCircuit]:
         # generate measurement circuits with metadata
```
Comment (ihincks) [inline qiskit/primitives/backend_estimator_v2.py]: I'm suggesting a name change and some docstring (not to be rendered, but for reference) to help readers of this file:

```suggestion
    def _create_measurement_circuits(
        self, circuit: QuantumCircuit, observable: PauliList, param_index: tuple[int, ...]
    ) -> list[QuantumCircuit]:
        """Generate a list of circuits sufficient to estimate each of the given Paulis.
        
        Paulis are divided into qubitwise-commuting subsets to reduce the total circuit count.
        Metadata is attached to circuits in order to remember what each one measures, and 
        where it belongs in the output.
        
        Args:
            circuit: The circuit of interest.
            observable: Which Pauli terms we would like to observe.
            param_index: Where to put the data we estimate (only passed to metadata).
        
        Returns:
            A list of circuits sufficient to estimate each of the given Paulis.
        """
```

File: qiskit/primitives/backend_estimator_v2.py [hunk 3]
```diff
@@ -178,30 +225,27 @@ def _run_pub(self, pub: EstimatorPub) -> PubResult:
         data_bin = data_bin_cls(evs=evs, stds=stds)
         return PubResult(data_bin, metadata={"target_precision": pub.precision})
 
-    def _calc_expval_paulis(
+    def _generate_circuits(
```
Comment (ihincks) [inline qiskit/primitives/backend_estimator_v2.py]:
```suggestion
    def _bind_and_add_measurements
```

File: qiskit/primitives/backend_estimator_v2.py [hunk 4]
```diff
@@ -178,30 +225,27 @@ def _run_pub(self, pub: EstimatorPub) -> PubResult:
...
         self,
         circuit: QuantumCircuit,
         parameter_values: BindingsArray,
         param_obs_map: dict[tuple[int, ...], set[str]],
-        shots: int,
-    ) -> dict[tuple[tuple[int, ...], str], tuple[float, float]]:
-        # generate circuits
+    ) -> list[QuantumCircuit]:
```
Comment (ihincks) [inline qiskit/primitives/backend_estimator_v2.py:268]: Suggesting a docstring to help future readers:

```suggestion
    ) -> list[QuantumCircuit]:
        """Bind the given circuit against each parameter value set, and add necessary measurements to each.
        
        Args:
            circuit: The (possibly parametric) circuit of interest.
            parameter_values: An array of parameter value sets that can be applied to the circuit.
            param_obs_map: A mapping from locations in ``parameter_values`` to a sets of 
                Pauli terms whose expectation values are required in those locations.
                
        Returns:
            A flat list of circuits sufficient to measure all Pauli terms in the ``param_obs_map`` values at the
            corresponding ``parameter_values`` location, where requisite book-keeping is stored as 
            circuit metadata.
        """
```

File: qiskit/primitives/backend_estimator_v2.py [hunk 5]
```diff
@@ -162,8 +201,16 @@ def _run_pub(self, pub: EstimatorPub) -> PubResult:
         for index in np.ndindex(*bc_param_ind.shape):
             param_index = bc_param_ind[index]
             param_obs_map[param_index].update(bc_obs[index].keys())
```
Comment (ihincks) [inline qiskit/primitives/backend_estimator_v2.py]:
```suggestion
            param_obs_map[param_index].update(bc_obs[index])
```

File: qiskit/primitives/backend_estimator_v2.py [hunk 6]
```diff
@@ -144,10 +145,48 @@ def _validate_pubs(self, pubs: list[EstimatorPub]):
...
+        results = []
+        start = 0
+        for pub, size, bc_param_ind, bc_obs in zip(pubs, sizes, bc_param_ind_list, bc_obs_list):
+            end = start + size
+            expval_map = self._calc_expval_map(counts[start:end], metadata[start:end])
+            start = end
+            results.append(self._postprocessing_pub(pub, expval_map, bc_param_ind, bc_obs, shots))
+        return results
+
+    def _preprocessing_pub(self, pub: EstimatorPub) -> list[QuantumCircuit]:
```
Comment (ihincks) [inline qiskit/primitives/backend_estimator_v2.py]:
```suggestion
    def _preprocess_pub(self, pub: EstimatorPub) -> tuple[list[QuantumCircuit], np.ndarray, np.ndarray]:
        """Converts a pub into a list of bound circuits necessary to estimate all its observables.
        
        The circuits contain metadata explaining which bindings array index they are with respect to, and 
        which measurement basis they are measuring.
        
        Args:
            pub: The pub to preprocess.
            
        Returns:
            The values ``(circuits, bc_param_ind, bc_obs)`` where ``circuits`` are the circuits to execute on the 
            backend, ``bc_param_ind`` are indices of the pub's bindings array and ``bc_obs`` is the observables
            array, both broadcast to the shape of the pub.
        """
```

File: qiskit/primitives/backend_estimator_v2.py [hunk 7]
```diff
@@ -162,8 +201,16 @@ def _run_pub(self, pub: EstimatorPub) -> PubResult:
...
             param_obs_map[param_index].update(bc_obs[index].keys())
-        expval_map = self._calc_expval_paulis(circuit, parameter_values, param_obs_map, shots)
 
+        return (
+            self._generate_circuits(circuit, parameter_values, param_obs_map),
+            bc_param_ind,
+            bc_obs,
+        )
+
+    def _postprocessing_pub(
```
Comment (ihincks) [inline qiskit/primitives/backend_estimator_v2.py]:
```suggestion
    def _postprocess_pub(
```

File: qiskit/primitives/backend_sampler_v2.py [hunk 8]
```diff
@@ -15,6 +15,7 @@
 from __future__ import annotations
```
Comment (ihincks) [inline qiskit/primitives/backend_sampler_v2.py:15]: Please consider analagous name changes in this file as suggested in `backend_estimator_v2.py`, for instance, switch to the imperative mood preprocessing->preprocess.

**Review by ihincks — COMMENTED** [review]
Thanks for this quick work. I've left a few comments, which I mostly made as I read through the file figuring out what was going on. It all makes sense to me, so LGTM!

File: qiskit/primitives/backend_sampler_v2.py (hunk 8 above)
Comment (t-imamichi) [reply qiskit/primitives/backend_sampler_v2.py:15]: Yes, I renamed `_postprocessing_pub` with `_postprocess_pub`.

**Comment by t-imamichi** [comment]
I refactored BackendEstimatorV2 by introducing an internal data class `_PreprocessedData`.

**Review by ihincks — APPROVED** [review]
LGTM

**Pull request merged**